p2p/libp2p/discovery/factory/peerDiscovererFactory.go

```diff
@@ -11,29 +11,40 @@ import (
 
 // NewPeerDiscoverer generates an implementation of PeerDiscoverer by parsing the p2pConfig struct
 // Errors if config is badly formatted
-func NewPeerDiscoverer(p2pConfig config.P2PConfig) (p2p.PeerDiscoverer, error) {
+func NewPeerDiscoverer(
+	context context.Context,
+	host discovery.ConnectableHost,
+	sharder p2p.CommonSharder,
+	p2pConfig config.P2PConfig,
+) (p2p.PeerDiscoverer, error) {
 	if p2pConfig.KadDhtPeerDiscovery.Enabled {
-		return createKadDhtPeerDiscoverer(p2pConfig)
+		return createKadDhtPeerDiscoverer(context, host, sharder, p2pConfig)
 	}
 
 	return discovery.NewNullDiscoverer(), nil
 }
 
-func createKadDhtPeerDiscoverer(p2pConfig config.P2PConfig) (p2p.PeerDiscoverer, error) {
+func createKadDhtPeerDiscoverer(
+	context context.Context,
+	host discovery.ConnectableHost,
+	sharder p2p.CommonSharder,
+	p2pConfig config.P2PConfig,
+) (p2p.PeerDiscoverer, error) {
 	arg := discovery.ArgKadDht{
+		Context:              context,
+		Host:                 host,
+		KddSharder:           sharder,
 		PeersRefreshInterval: time.Second * time.Duration(p2pConfig.KadDhtPeerDiscovery.RefreshIntervalInSec),
 		RandezVous:           p2pConfig.KadDhtPeerDiscovery.RandezVous,
 		InitialPeersList:     p2pConfig.KadDhtPeerDiscovery.InitialPeerList,
 		BucketSize:           p2pConfig.KadDhtPeerDiscovery.BucketSize,
 		RoutingTableRefresh:  time.Second * time.Duration(p2pConfig.KadDhtPeerDiscovery.RefreshIntervalInSec),
 	}
 
-	switch p2pConfig.KadDhtPeerDiscovery.Type {
-	case config.KadDhtVariantPrioBits:
+	switch p2pConfig.Sharding.Type {
+	case p2p.SharderVariantPrioBits:
 		return discovery.NewKadDhtPeerDiscoverer(arg)
-	case config.KadDhtVariantWithLists:
-		return discovery.NewContinuousKadDhtDiscoverer(arg)
 	default:
```

Comment:
add the p2p.SharderVariantList as a case, and leave the default with error.

Reply:
Can not do that because I have 3 more implementations that are compatible with this one.
True, I can write 
```GO
case p2p.SharderVariantWithLists, p2p.NoSharderWithLists, p2p.DisabledSharder:
```
but I do not prefer this variant as I need to modify this file each time I will create a new sharder implementation. And we might forget to add this check.